FHEM/00_SIGNALduino.pm

```diff
@@ -1477,8 +1477,67 @@ sub SIGNALduino_Write {
 	    $sum	= 12;
   	}
     $msg = $hash->{protocolObject}->PreparingSend_FS20_FHT($id, $sum, $msg);
-  } elsif($fn eq 'k') {   # KOPP_FC
-		$hash->{logMethod}->($name, 4, "$name: Write, cmd $fn sending KOPP_FC");
+  } elsif($fn eq 'k') {   # KOPP_FC   (no outsourcing  in SD_Protocols.pm due to loop and set hash values)
+    $hash->{logMethod}->($name, 4, "$name: Write, cmd $fn sending KOPP_FC");
+    $fn='raw';
+
+    # https://wiki.fhem.de/wiki/Kopp_Allgemein | https://github.com/heliflieger/a-culfw/blob/master/culfw/clib/kopp-fc.c
+    #kr07C2AD1A30CC0F0328
+    #||  ||||  ||    ++-------- Transmitter Code 2
+    #||  ||||  ++-------------- Keycode
+    #||  ++++------------------ Transmitter Code 1
+    #++------------------------ kr wird von der culfw bei Empfang einer Kopp Botschaft als Kennung gesendet
+
+    # $message = "s"
+    #  . $keycodehex
+    #  . $hash->{TRANSMITTERCODE1}
+    #  . $hash->{TRANSMITTERCODE2}
+    #  . $hash->{TIMEOUT}
+    #  . "N";                       # N for do not print messages (FHEM will write error messages to log files if CCD/CUL sends status info
+
+    my $Keycode = substr($msg,1,2);
+    my $TransCode1 = substr($msg,3,4);
+    my $TransCode2 = substr($msg,7,2);
+    my $blkck = 0xAA;
+    my $d;
+
+    ### The device to be sent stores something in own hash. Search for names to access them ###
+    #### The variant with the loop does not require any adjustment in the original Kopp module. ####
+    my $KOPPname;
+    KOPPname:
+    foreach my $d (keys %defs) {
+      if(defined($defs{$d}) && $defs{$d}{TYPE} eq 'KOPP_FC') {
+        if ( (exists $defs{$d}->{TRANSMITTERCODE1} && $defs{$d}->{TRANSMITTERCODE1} eq $TransCode1) && 
+              (exists $defs{$d}->{TRANSMITTERCODE2} && $defs{$d}->{TRANSMITTERCODE2} eq $TransCode2) && 
+                (exists $defs{$d}->{KEYCODE} && $defs{$d}->{KEYCODE} eq $Keycode)
+              ) {
+          $hash->{logMethod}->($name, 5, "$name: Write, PreparingSend KOPP_FC found device with name $d");
+          $KOPPname = $d;
+        }
+        last KOPPname;
+      }
+    }
+
+    ## Internals blkctr initialize if not available
+    if (!exists($defs{$KOPPname}->{blkctr})) {
+      $defs{$KOPPname}->{blkctr} = 0;
+      $hash->{logMethod}->($name, 5, "$name: Write, PreparingSend KOPP_FC set blkctr in hash $KOPPname");
+    ## Internals blkctr increases with each send
+    }
+
+    my $dmsg = '07' . $TransCode1 . sprintf("%02x",$defs{$KOPPname}->{blkctr}) . $Keycode . 'CC0F' . $TransCode2;
+
+    ## checksum to calculate
+    for (my $i = 0; $i < 8; $i++) {
+      $d = hex(substr($dmsg,$i*2,2));
+      $blkck ^= $d;
+    }
+
+    $dmsg.= sprintf("%02x",$blkck) . '000000000000;';
```

Comment:
Die 0 kann man auch gleich einbauen oder auch als String übergeben. Ganz nach belieben.
`$dmsg.=sprintf("%02x000000000000",$blkck);`

Reply:
Die Art mit dem String finde ich auf den ersten Blick sofort einläuchtender. Somit weiß ich das es gefolgt nach dem sprintf kommt fest.

Reply:
Ich würd das dann gleich alles in sprintf zusammenbauen...
Was hälst Du denn davon?

`sprintf("%02x%s",$blkck,'000000000000;'); `


Oder noch besser, weil wir hier nicht so viele 0 zählen müssen um zu erkennen dass es 12 sind.
`sprintf("%02x%012d%s",43455,undef,';'); `

Reply:
Diese Zeile `sprintf("%02x%012d%s",43455,undef,';'); ` kann ich nicht richtig deuten.
Woher kommen die 43455? Meinst du damit `$blkck` ?

Reply:
43455  ist quasi $blkck     
%02x = zwei zweistellige hexwerte mit führender 0
%012d  = 12 führende nullen
%s = ein string

Reply:
Werde ich mal umstellen und testen.